File: treebolicGlue/src/main/java/treebolic/glue/Animator.java
/*
 * Copyright (c) 2019-2023. Bernard Bou
 */

package treebolic.glue;

import android.animation.ValueAnimator;
import android.animation.ValueAnimator.AnimatorUpdateListener;
import android.view.animation.LinearInterpolator;

import androidx.annotation.NonNull;

/**
 * Animator
 *
 * @author Bernard Bou
 */
public class Animator implements treebolic.glue.iface.Animator<ActionListener>, AnimatorUpdateListener, android.animation.Animator.AnimatorListener
{
	// static private final String TAG = "Animator";

	/**
	 * Animation time slice
	 */
	static private final int ANIMATIONTIMESLICE = 250;

	/**
	 * Animation listener
	 */
	@SuppressWarnings("InstanceVariableOfConcreteClass")
	private ActionListener listener;

	/**
	 * Android animator
	 */
	private ValueAnimator animator;

	/**
	 * Android last step
	 */
	private int lastStep;

	/**
	 * Constructor
	 */
	public Animator()
	{
		ValueAnimator.setFrameDelay(Animator.ANIMATIONTIMESLICE);
		// Log.d(Animator.TAG, "animate frame delay set=" + ANIMATIONTIMESLICE + " get=" + ValueAnimator.getFrameDelay());
	}

	@SuppressWarnings({"UnusedReturnValue", "SameReturnValue"})
	@Override
	public boolean run(final ActionListener listener, final int steps, final int startDelay)
	{
		// Log.d(Animator.TAG, "animate steps " + steps);
		this.lastStep = steps - 1;
		this.listener = listener;
		this.animator = ValueAnimator.ofInt(0, this.lastStep);
		this.animator.setRepeatCount(0);
		this.animator.setDuration(1000);
		this.animator.setStartDelay(startDelay);
		this.animator.setInterpolator(new LinearInterpolator());
		this.animator.addUpdateListener(this);
		this.animator.addListener(this);
		this.animator.start();
		return true;
	}

	@Override
	public boolean isRunning()
	{
		return this.animator.isRunning();
	}

	@SuppressWarnings("boxing")
	@Override
	public void onAnimationUpdate(@NonNull final ValueAnimator animator)
	{
		int step = (int) animator.getAnimatedValue();
		this.listener.onAction(step);
		// Log.d(Animator.TAG, "animate update value=" + animator.getAnimatedValue());
	}

	@SuppressWarnings("EmptyMethod")
	@Override
	public void onAnimationStart(@NonNull android.animation.Animator animator)
	{
		// Log.d(Animator.TAG, "animate start " + Animator.frameCount + " value=" + ((ValueAnimator) animator).getAnimatedValue());
	}

	@SuppressWarnings("boxing")
	@Override
	public void onAnimationEnd(@NonNull android.animation.Animator animator)
	{
		this.listener.onAction(this.lastStep);
		// Log.d(Animator.TAG, "animate end value=" + this.lastStep);
	}

	@SuppressWarnings("boxing")
	@Override
	public void onAnimationCancel(@NonNull android.animation.Animator animator)
	{
		this.listener.onAction(this.lastStep);
		// Log.d(Animator.TAG, "animate cancel value=" + ((ValueAnimator) animator).getAnimatedValue());
	}

	@SuppressWarnings("EmptyMethod")
	@Override
	public void onAnimationRepeat(@NonNull android.animation.Animator animator)
	{
		// Log.d(Animator.TAG, "animate repeat " + Animator.frameCount + " value=" + ((ValueAnimator) animator).getAnimatedValue());
	}
}

File: treebolicGlue/src/main/java/treebolic/glue/Animator.kt
/*
 * Copyright (c) 2019-2023. Bernard Bou
 */
package treebolic.glue

import android.animation.ValueAnimator
import android.animation.ValueAnimator.AnimatorUpdateListener
import android.view.animation.LinearInterpolator
import treebolic.glue.iface.Animator

/**
 * Animator
 *
 * @author Bernard Bou
 */
class Animator : Animator<ActionListener?>, AnimatorUpdateListener, android.animation.Animator.AnimatorListener {

    /**
     * Animation listener
     */
    private lateinit var listener: ActionListener

    /**
     * Android animator
     */
    private lateinit var animator: ValueAnimator

    /**
     * Android last step
     */
    private var lastStep = 0

    /**
     * Constructor
     */
    init {
        ValueAnimator.setFrameDelay(ANIMATIONTIMESLICE.toLong())
    }

    override fun run(listener0: ActionListener?, steps: Int, startDelay: Int): Boolean {
        lastStep = steps - 1
        listener = listener0!!
        animator = ValueAnimator.ofInt(0, lastStep)
        animator.repeatCount = 0
        animator.setDuration(1000)
        animator.setStartDelay(startDelay.toLong())
        animator.interpolator = LinearInterpolator()
        animator.addUpdateListener(this)
        animator.addListener(this)
        animator.start()
        return true
    }

    override fun isRunning(): Boolean {
        return animator.isRunning
    }

    override fun onAnimationUpdate(animator: ValueAnimator) {
        val step = animator.animatedValue as Int
        listener.onAction(step)
    }

    override fun onAnimationStart(animator: android.animation.Animator) {
    }

    override fun onAnimationEnd(animator: android.animation.Animator) {
        listener.onAction(lastStep)
    }

    override fun onAnimationCancel(animator: android.animation.Animator) {
        listener.onAction(lastStep)
    }

    override fun onAnimationRepeat(animator: android.animation.Animator) {
    }

    companion object {

        /**
         * Animation time slice
         */
        private const val ANIMATIONTIMESLICE = 250
    }
}
